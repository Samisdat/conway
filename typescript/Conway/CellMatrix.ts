import {Cell} from './cell';
import {Position} from './position';

interface CellMatrixInterface {
    [index: string]: Cell;
}

export class CellMatrix {

    private matrix: CellMatrixInterface = {};

    public get(position: Position): Cell | undefined {

        if (undefined !== this.matrix[position.toString()]) {
            return this.matrix[position.toString()];
        }

        return undefined;


    }

    public add(cell: Cell): void {

        if (undefined !== this.matrix[cell.position.toString()]) {
            console.log('argh');
        }

        this.matrix[cell.position.toString()] = cell;

    }

    public remove(cell: Cell): void {

        if (undefined !== this.matrix[cell.position.toString()]) {
            delete this.matrix[cell.position.toString()];
        }

    }

    public all(): Cell[] {

        const cells: Cell[] = [];

        for (const index in this.matrix) {

            cells.push(this.matrix[index]);

        }

        return cells;

    }


}
