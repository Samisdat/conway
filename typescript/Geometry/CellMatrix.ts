import {Position} from '@Conway/Geometry/Position';

interface CellMatrixInterface {
    [index: string]: Position;
}

interface Dimension {
    width: number;
    height: number;
}

export class CellMatrix {

    private matrix: CellMatrixInterface = {};

    public has(position: Position): boolean {

        return (undefined !== this.matrix[position.toString()]);

    }

    public add(position: Position): void {

        if (true === this.has(position)) {
            return;
        }

        this.matrix[position.toString()] = position;

    }

    public remove(position: Position): void {

        if (false === this.has(position)) {
            return;
        }

        delete this.matrix[position.toString()];

    }

    public all(): Position[] {

        const positions: Position[] = [];

        for (const index in this.matrix) {

            positions.push(this.matrix[index]);

        }

        return positions;

    }

    private getDimension(): Dimension {

        const dimension: Dimension = {
            width: 0,
            height: 0
        }

        let left = 0;
        let right = 0;

        let top = 0;
        let bottom = 0;

        for (const position of this.all()) {

            if (position.x < left) {
                left = position.x;
            }

            if (position.x > right) {
                right = position.x;
            }

            if (position.y < bottom) {
                bottom = position.y;
            }

            if (position.y > top) {
                top = position.y;
            }

        }

        if (0 > left && 0 < right) {
            dimension.width = right - left;
        }
        else{
            dimension.width = right - left;
        }

        if (0 > bottom && 0 < top) {
            dimension.height = top - bottom;
        }
        else{
            dimension.height = top - bottom;
        }

        return dimension;

    }

    public width(): number {

        return this.getDimension().width;

    }

    public height(): number {

        return this.getDimension().height;

    }

}
