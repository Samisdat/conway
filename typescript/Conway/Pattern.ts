import {CellMatrix} from '@Conway/Conway/CellMatrix';
import {Position} from '@Conway/Conway/Position';

export class Pattern {

    private name: String;

    private matrix: CellMatrix;

    constructor(name: String, matrix: CellMatrix) {

        this.name = name;

        this.matrix = matrix;

    }

    public getMatrix(): CellMatrix {

        return this.matrix;
    }

    public getWidth(): number {
        return this.matrix.width();
    }

    public getHeight(): number {
        return this.matrix.height();
    }

    public mirrorVertical(): void {

        const mirror = new CellMatrix();

        for (const position of this.matrix.all()) {

            mirror.add(
                new Position(
                    position.x,
                    -1 * position.y
                )
            );

        }

        this.matrix = mirror;

    }

    public mirrorHorizontal(): void {

        const mirror = new CellMatrix();

        for (const position of this.matrix.all()) {

            mirror.add(
                new Position(
                    -1 * position.x,
                    position.y
                )
            );

        }

        this.matrix = mirror;

    }

}
