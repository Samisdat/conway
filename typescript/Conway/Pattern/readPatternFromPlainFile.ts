import {CellMatrix} from '@Conway/Conway/CellMatrix';
import {Position} from '@Conway/Conway/Position';
import {Pattern} from '@Conway/Conway/Pattern';

export const readPatternFromPlainFile = (patternPlainText: string): Pattern => {

    let name: string | undefined = undefined;
    const matrix = new CellMatrix();

    const patternPows: string[] = patternPlainText.split('\n');

    const isComment = new RegExp('^!(.*?)');
    const isMeta = new RegExp('^!(.*?)\:(.*?)$');

    let y = 0;
    let x = 0;

    for (let patternPow of patternPows) {

        patternPow = patternPow.trim();

        if (true === isComment.test(patternPow)) {

            if (true === isMeta.test(patternPow)) {

                const meta = isMeta.exec(patternPow);

                if (null === meta) {
                    continue;
                }

                if ('Name' === meta[1]) {
                    name = meta[2].trim();
                }

            }
            continue;
        }

        const cols: string[] = patternPow.trim().split('');

        x = 0;

        for (const col of cols) {

            if ('O' === col) {

                matrix.add(new Position(x, y));

            }

            x += 1;
        }

        y += 1;

    }

    if (undefined === name) {
        name = 'unnamed';
    }

    const pattern = new Pattern(name, matrix);
    //pattern.mirrorVertical();

    return pattern;

}

