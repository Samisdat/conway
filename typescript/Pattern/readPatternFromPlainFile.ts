import {Position} from '@Conway/Geometry/Position';
import {Pattern} from '@Conway/Pattern/Pattern';
import {Matrix} from '@Conway/Geometry/Matrix';

export const readPatternFromPlainFile = (patternPlainText: string): Pattern => {

    let name: string | undefined = undefined;
    const matrix = new Matrix();

    const patternPows: string[] = patternPlainText.split('\n').reverse();

    const isComment = new RegExp('^!(.*?)');
    const isMeta = new RegExp('^!(.*?)\:(.*?)$');

    let y = 0;
    let x = 0;

    for (let patternPow of patternPows) {

        patternPow = patternPow.trim();

        if (true === isComment.test(patternPow)) {

            if (true === isMeta.test(patternPow)) {

                const meta = isMeta.exec(patternPow) as RegExpMatchArray;

                if ('Name' === meta[1]) {
                    name = meta[2].trim();
                }

            }
            continue;
        }

        const cols: string[] = patternPow.trim().split('');

        x = 0;

        for (const col of cols) {

            if ('O' === col) {

                matrix.add(new Position(x, y));

            }

            x += 1;
        }

        y += 1;

    }

    if (undefined === name) {
        name = 'unnamed';
    }

    const pattern = new Pattern(name, matrix);

    return pattern;

};

