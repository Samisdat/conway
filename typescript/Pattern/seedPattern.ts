import {CellMatrix} from '@Conway/Geometry/CellMatrix';
import {Pattern} from '@Conway/Pattern/Pattern';
import {Position} from '@Conway/Geometry/Position';

export const seedPattern = (matrix: CellMatrix, pattern: Pattern, moveBy = new Position(0, 0)): void => {

    const positions = pattern.getMatrix().all();

    const offSet = new Position(
        -1 * Math.floor(pattern.getMatrix().width() / 2),
        -1 * Math.floor(pattern.getMatrix().height() / 2)
    ).move(moveBy);

    for (const position of positions) {

        matrix.add(position.move(moveBy));

    }

}
