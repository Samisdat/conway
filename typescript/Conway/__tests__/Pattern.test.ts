import {Pattern} from '../Pattern';
import {Position} from '../Position';
import {CellMatrix} from '../CellMatrix';
import serializer from '../../../jest-serialize-conway';

describe('Pattern', () => {

    beforeEach(function () {
        expect.addSnapshotSerializer(serializer);
    });

    test('can be created', () => {

        const cellMatrix = new CellMatrix();
        cellMatrix.add(
            new Position(-1, -1)
        );

        let pattern = new Pattern(
            'Scottish',
            cellMatrix
        );

        expect(pattern).toBeInstanceOf(Pattern);

    });

    test('get width and height', () => {

        const cellMatrix = new CellMatrix();
        cellMatrix.add(
            new Position(-1, -1)
        );
        cellMatrix.add(
            new Position(1, 2)
        );

        let pattern = new Pattern(
            'Scottish',
            cellMatrix
        );

        expect(pattern.getWidth()).toBe(3);
        expect(pattern.getHeight()).toBe(4);

        expect(pattern.getMatrix()).toMatchSnapshot();

    });

    test('mirrorHorizontal simple', () => {

        const cellMatrix = new CellMatrix();
        cellMatrix.add(
            new Position(-1, 1)
        );
        cellMatrix.add(
            new Position(-1, 0)
        );
        cellMatrix.add(
            new Position(-1, -1)
        );

        let pattern = new Pattern(
            'Scottish',
            cellMatrix
        );

        expect(pattern.getWidth()).toBe(1);
        expect(pattern.getHeight()).toBe(3);

        expect(pattern.getMatrix()).toMatchSnapshot();

        pattern.mirrorHorizontal();

        expect(pattern.getMatrix()).toMatchSnapshot();

    });

    test('mirrorVertical simple', () => {

        const cellMatrix = new CellMatrix();
        cellMatrix.add(
            new Position(-1, 1)
        );
        cellMatrix.add(
            new Position(0, 1)
        );
        cellMatrix.add(
            new Position(1, 1)
        );

        let pattern = new Pattern(
            'Scottish',
            cellMatrix
        );

        expect(pattern.getWidth()).toBe(3);
        expect(pattern.getHeight()).toBe(1);

        expect(pattern.getMatrix()).toMatchSnapshot();

        pattern.mirrorVertical();

        expect(pattern.getMatrix()).toMatchSnapshot();


    });

});
