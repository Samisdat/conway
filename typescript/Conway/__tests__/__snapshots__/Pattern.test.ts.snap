// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pattern get width and height 1`] = `
OOX
OOO
OOO
XOO
`;

exports[`Pattern mirrorHorizontal simple 1`] = `
XO
XO
XO
`;

exports[`Pattern mirrorHorizontal simple 2`] = `
OX
OX
OX
`;

exports[`Pattern mirrorVertical simple 1`] = `
XXX
OOO
`;

exports[`Pattern mirrorVertical simple 2`] = `
OOO
XXX
`;
